File: src/main/java/org/dbtools/query/shared/filter/CompareFilter.java
package org.dbtools.query.shared.filter;

import org.dbtools.query.shared.QueryBuilder;
import org.dbtools.query.shared.CompareType;

import javax.annotation.Nonnull;

public class CompareFilter extends Filter {

    protected String field;
    private CompareType compareType;
    protected Object value;

    public static CompareFilter create(String field, Object value) {
        CompareFilter filter = new CompareFilter();
        filter.filter = CompareFilter.newInstance(field, CompareType.EQUAL, value);
        return filter;
    }

    public static CompareFilter create(String field, CompareType compareType, Object value) {
        CompareFilter filter = new CompareFilter();
        filter.filter = CompareFilter.newInstance(field, compareType, value);
        return filter;
    }

    public static CompareFilter create(String field, CompareType compareType) {
        CompareFilter filter = new CompareFilter();
        filter.filter = CompareFilter.newInstance(field, compareType);
        return filter;
    }

    private static CompareFilter newInstance(String field, CompareType compareType, Object value) {
        CompareFilter filter;
        switch (compareType) {
            case LIKE:
                filter = LikeFilter.create(field, value, false);
                break;
            case LIKE_IGNORECASE:
                filter = LikeFilter.create(field, value, true);
                break;
            case IN:
                filter = InFilter.create(field, true, value);
                break;
            case NOT_IN:
                filter = InFilter.create(field, false, value);
                break;
            case IS_NULL:
                filter = NullFilter.create(field, true);
                break;
            case NOT_NULL:
                filter = NullFilter.create(field, false);
                break;
            default:
                filter = new CompareFilter(field, compareType, value);
        }
        return filter;
    }

    private static CompareFilter newInstance(String field, CompareType compareType) {
        CompareFilter filter;
        switch (compareType) {
            case IS_NULL:
            case NOT_NULL:
                filter = CompareFilter.newInstance(field, compareType, null);
                break;
            default:
                throw new IllegalArgumentException("Illegal 1 argument compare " + compareType.toString());
        }
        return filter;
    }

    protected CompareFilter() {
    }

    protected CompareFilter(String field, CompareType compareType, Object value) {
        this.field = field;
        this.compareType = compareType;
        this.value = value;
    }

    @Override
    public String build(@Nonnull QueryBuilder queryBuilder) {
        StringBuilder builder = new StringBuilder(field);
        switch (compareType) {
            case EQUAL:
                builder.append(" = ");
                break;
            case NOT_EQUAL:
                builder.append(" != ");
                break;
            case LESSTHAN:
                builder.append(" < ");
                break;
            case GREATERTHAN:
                builder.append(" > ");
                break;
            case LESSTHAN_EQUAL:
                builder.append(" <= ");
                break;
            case GREATERTHAN_EQUAL:
                builder.append(" >= ");
                break;
            default:
                throw new IllegalStateException("Invalid QueryCompareType: " + compareType);
        }
        builder.append(queryBuilder.formatValue(value));
        return builder.toString();
    }

    public CompareFilter and(String field, Object value) {
        and(CompareFilter.newInstance(field, CompareType.EQUAL, value));
        return this;
    }

    public CompareFilter and(String field, CompareType compareType, Object value) {
        and(CompareFilter.newInstance(field, compareType, value));
        return this;
    }

    public CompareFilter and(String field, CompareType compareType) {
        and(CompareFilter.newInstance(field, compareType));
        return this;
    }

    public CompareFilter or(String field, Object value) {
        or(CompareFilter.newInstance(field, CompareType.EQUAL, value));
        return this;
    }

    public CompareFilter or(String field, CompareType compareType, Object value) {
        or(CompareFilter.newInstance(field, compareType, value));
        return this;
    }

    public CompareFilter or(String field, CompareType compareType) {
        or(CompareFilter.newInstance(field, compareType));
        return this;
    }


    @Override
    public CompareFilter clone() {
        CompareFilter clone = (CompareFilter) super.clone();
        clone.field = this.field;
        clone.compareType = this.compareType;
        clone.value = this.value;
        return clone;
    }
}

File: dbtools-query/src/commonMain/kotlin/org/dbtools/query/shared/filter/CompareFilter.kt
package org.dbtools.query.shared.filter

import org.dbtools.query.shared.CompareType
import org.dbtools.query.shared.QueryBuilder

open class CompareFilter(
    protected var field: String? = null,
    protected var compareType: CompareType? = null,
    protected var value: Any? = null
) : Filter() {
    override fun build(queryBuilder: QueryBuilder): String = buildString {
        append(field)
        when (compareType) {
            CompareType.EQUAL -> append(" = ")
            CompareType.NOT_EQUAL -> append(" != ")
            CompareType.LESSTHAN -> append(" < ")
            CompareType.GREATERTHAN -> append(" > ")
            CompareType.LESSTHAN_EQUAL -> append(" <= ")
            CompareType.GREATERTHAN_EQUAL -> append(" >= ")
            CompareType.NOT_NULL -> append(" NOT NULL")
            CompareType.IS_NULL -> append(" IS NULL")
            else -> error("Invalid QueryCompareType: $compareType")
        }

        if (value != null) {
            append(queryBuilder.formatValue(value))
        }
    }

    open fun and(field: String, value: Any): CompareFilter {
        and(newInstance(field, CompareType.EQUAL, value))
        return this
    }

    fun and(field: String, compareType: CompareType, value: Any): CompareFilter {
        and(newInstance(field, compareType, value))
        return this
    }

    fun and(field: String, compareType: CompareType): CompareFilter {
        and(newInstance(field, compareType))
        return this
    }

    open fun or(field: String, value: Any): CompareFilter {
        or(newInstance(field, CompareType.EQUAL, value))
        return this
    }

    fun or(field: String, compareType: CompareType, value: Any): CompareFilter {
        or(newInstance(field, compareType, value))
        return this
    }

    fun or(field: String, compareType: CompareType): CompareFilter {
        or(newInstance(field, compareType))
        return this
    }

    companion object {
        fun create(field: String, value: Any): CompareFilter {
            val filter = CompareFilter()
            filter.filter = newInstance(field, CompareType.EQUAL, value)
            return filter
        }

        fun create(field: String, compareType: CompareType, value: Any): CompareFilter {
            val filter = CompareFilter()
            filter.filter = newInstance(field, compareType, value)
            return filter
        }

        fun create(field: String, compareType: CompareType): CompareFilter {
            val filter = CompareFilter()
            filter.filter = newInstance(field, compareType)
            return filter
        }

        private fun newInstance(field: String, compareType: CompareType, value: Any): CompareFilter {
            val filter = when (compareType) {
                CompareType.LIKE -> LikeFilter.create(field, value, false)
                CompareType.LIKE_IGNORECASE -> LikeFilter.create(field, value, true)
                CompareType.IN -> InFilter.create(field, true, value)
                CompareType.NOT_IN -> InFilter.create(field, false, value)
                CompareType.IS_NULL -> NullFilter.create(field, true)
                CompareType.NOT_NULL -> NullFilter.create(field, false)
                else -> CompareFilter(field, compareType, value)
            }
            return filter
        }

        private fun newInstance(field: String, compareType: CompareType): CompareFilter {
            return when (compareType) {
//                CompareType.IS_NULL, CompareType.NOT_NULL -> newInstance(field, compareType, null)
                CompareType.IS_NULL -> NullFilter.create(field, true)
                CompareType.NOT_NULL -> NullFilter.create(field, false)
                else -> throw IllegalArgumentException("Illegal 1 argument compare $compareType")
            }
        }
    }
}
